System/Console/ANSI/Unix.hs

```diff
@@ -42,3 +56,45 @@ hHideCursor h = hPutStr h hideCursorCode
 hShowCursor h = hPutStr h showCursorCode
 
 hSetTitle h title = hPutStr h $ setTitleCode title
+
+-- getReportedCursorPosition :: IO String
+-- (See Common-Include.hs for Haddock documentation)
+getReportedCursorPosition = do
+    echo <- hGetEcho stdin -- Preserve existing echo
+    hSetEcho stdin False   -- Turn echo off
+    cp <- get
+    hSetEcho stdin echo    -- Restore echo
+    return cp
+  where
+    get = do
+        c <- getChar
+        if c == '\ESC'
+            then get' [c]
+            else return [c] -- If the first character is not the expected
+                            -- \ESC then give up. This provides a modicom of
+                            -- protection against unexpected data in the
+                            -- input stream.
+    get' s = do
+        c <- getChar
+        if c == 'R'
+            then return (s ++ [c])
+            else get' (s ++ [c]) -- Continue until the expected 'R' character is
```

Comment:
This has quadratic complexity. Our input is probably short, but let's accumulate the list in the reverse just to be safe...

Reply:
Now accumulated in reverse, and then `reverse`-d, and a comment added to explain why.